### PR TITLE
[Python] Move existing dependency installation and python command invocation to a Toolchain interface

### PR DESCRIPTION
**Pull request by julienp**
# Description

This PR refactors the existing Python dependency installation and command running code to use the `Toolchain` interface. This will make it possible to swap out the default Pip based toolchain for a Poetry based toolchain.

Fixes https://github.com/pulumi/pulumi/issues/16285
Ref https://github.com/pulumi/pulumi/issues/15937

## Checklist

- [x] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [x] I have formatted my code using `gofumpt`

### CODE REVIEW

File: sdk/python/toolchain/pip.go [hunk 1]
```diff
@@ -26,13 +28,125 @@ import (
...
+func newPip(root, virtualenv string) (*pip, error) {
+	virtualenvPath := resolveVirtualEnvironmentPath(root, virtualenv)
+	logging.V(9).Infof("Python toolchain: using pip at %s", virtualenvPath)
+	return &pip{virtualenvPath, virtualenv, root}, nil
+}
+
+func (p *pip) InstallDependencies(ctx context.Context, cwd string, showOutput bool,
+	infoWriter io.Writer, errorWriter io.Writer,
+) error {
+	return InstallDependencies(
```
Comment (julienp) [inline sdk/python/toolchain/pip.go:59]: The existing `InstallDependencies` can be inlined here, and there's more tidying up that can be done in this file, but the diff is already fairly large, so I am leaving that for a later PR.

File: sdk/go/auto/errors_test.go [hunk 2]
```diff
@@ -416,7 +426,11 @@ func TestRuntimeErrorPython(t *testing.T) {
 	}
 
 	// install Pulumi Python SDK from the current source tree, -e means no-copy, ref directly
-	pyCmd := python.VirtualEnvCommand(filepath.Join(pDir, "venv"), "python", "-m", "pip", "install", "-e", pySDK)
+	pyCmd, err := tc.Command(context.Background(), "-m", "pip", "install", "-e", pySDK)
```
Comment (tgummerer) [inline sdk/go/auto/errors_test.go:429]: I see `-m` being passed in to `tc.Command` as first argument everywhere.  Makes me wonder if we should just have `tc.Command` prepend that?  Not sure if there's any python toolchains out there that don't require that, but I could imagine someone e.g. wanting a "System" toolchain that doesn't rely on a venv at all, that would be made possible if we don't pass that as first argument everywhere?  Or is there another reason for this?

File: sdk/python/cmd/pulumi-language-python/main.go [hunk 3]
```diff
@@ -97,6 +97,7 @@ func main() {
 	flag.String("virtualenv", "", "[obsolete] Virtual environment path to use")
 	flag.String("root", "", "[obsolete] Project root path to use")
 	flag.String("typechecker", "", "[obsolete] Use a typechecker to type check")
+	flag.String("toolchain", "pip", "[obsolete] Select the package manager to use for dependency management.")
```
Comment (tgummerer) [inline sdk/python/cmd/pulumi-language-python/main.go:100]: Curious, why are we adding a new option that's obsolete here?

File: sdk/python/cmd/pulumi-language-python/main_test.go [hunk 4]
```diff
@@ -234,3 +254,24 @@ func TestDeterminePulumiPackages(t *testing.T) {
...
+func runPythonCommand(t *testing.T, virtualenv, cwd string, arg ...string) ([]byte, error) {
+	t.Helper()
+
+	tc, err := toolchain.ResolveToolchain(toolchain.PythonOptions{
+		Toolchain:  toolchain.Pip,
+		Root:       cwd,
+		Virtualenv: virtualenv,
+	})
+	require.NoError(t, err)
+	cmd, err := tc.Command(context.Background(), arg...)
```
Comment (tgummerer) [inline sdk/python/cmd/pulumi-language-python/main_test.go:267]: We used to pass `ctx` into `runPythonCommand`.  Why did we change that and use `context.Background()` here instead?

File: sdk/python/toolchain/pip.go [hunk 5]
```diff
@@ -26,13 +28,125 @@ import (
...
+	virtualenvOption string
+	// The root directory of the project.
+	root string
+}
+
+var _ Toolchain = &pip{}
+
+func newPip(root, virtualenv string) (*pip, error) {
+	virtualenvPath := resolveVirtualEnvironmentPath(root, virtualenv)
+	logging.V(9).Infof("Python toolchain: using pip at %s", virtualenvPath)
```
Comment (tgummerer) [inline sdk/python/toolchain/pip.go:52]: This is very much besides the PR, and I feel like something I should know already, but do we have any docs on log levels anywhere?  I.e. what should be logged at 9 vs. 5 vs. whatever?

File: sdk/python/toolchain/pip.go (hunk 1 above)
Comment (tgummerer) [reply sdk/python/toolchain/pip.go:59]: Makes sense, this PR is quite large to get through already :)

File: sdk/go/auto/errors_test.go (hunk 2 above)
Comment (julienp) [reply sdk/go/auto/errors_test.go:429]: The "system" toolchain can be used by not passing a `Virtualenv` argument. We don't pass `-m` when running the user's pulumi program for example. I think there might be some cases in the tests as well.

Maybe we can have the existing `toolchain.Command` and `toolchain.ModuleCommand` for when we use `-m` https://docs.python.org/3/using/cmdline.html#cmdoption-m

File: sdk/python/cmd/pulumi-language-python/main.go (hunk 3 above)
Comment (julienp) [reply sdk/python/cmd/pulumi-language-python/main.go:100]: I had previously wondered about that too, Fraser explained here https://github.com/pulumi/pulumi/pull/15725#discussion_r1540785931

> Passing runtime flags as command line parameters to the language hosts is a deprecated form of passing them. You'll note that main doesn't do anything with these flags.
BUT the engine still sends these flags to language hosts for the time being as a back-compat workaround till all the hosts are updated, so the parser needs to recognize them so it doesn't fail on startup with "unrecognized argument".

File: sdk/python/cmd/pulumi-language-python/main_test.go (hunk 4 above)
Comment (julienp) [reply sdk/python/cmd/pulumi-language-python/main_test.go:267]: This is a test helper only, and all the uses of it pass in `context.Background()`.

File: sdk/python/toolchain/pip.go (hunk 5 above)
Comment (julienp) [reply sdk/python/toolchain/pip.go:52]: Yeah I don't know 🤷

File: sdk/go/auto/errors_test.go (hunk 2 above)
Comment (tgummerer) [reply sdk/go/auto/errors_test.go:429]: Ah that makes sense.  I do think a `ModuleCommand` would be nice to have.

Comment (julienp) [reply sdk/go/auto/errors_test.go]: I added `ModuleCommand` here https://github.com/pulumi/pulumi/pull/16243/commits/62de2b049261b96a2b2e91f5ced7d139debb2339

**Review by tgummerer — APPROVED** [review]
LGTM!

**Pull request merged**